### PR TITLE
Add libtheora

### PR DESCRIPTION
**Pull request by vlj**
There's no CMake port on the theora git repo, I wrote one based one the vcproj files in win32.

### CODE REVIEW

File: ports/libtheora/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+add_definitions(-D_USRDLL -DLIBTHEORA_EXPORTS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_BIND_TO_CURRENT_CRT_VERSION -DWIN32 -DNDEBUG -D_WINDOWS)
+
+add_library(theora ${SOURCES})
+target_link_libraries(theora ${LIBOGG_PATH})
+
+install(FILES ${HEADERS} DESTINATION include/theora)
+
+install(TARGETS theora
+  RUNTIME DESTINATION bin
+  LIBRARY DESTINATION lib
```
Comment (Orvid) [inline ports/libtheora/CMakeLists.txt]: Since you have control over it, you want this as `bin`, as this is referring to `.dll`s.

File: ports/libtheora/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+"lib/quant.c"
+"lib/state.c"
+"lib/x86_vc/mmxfrag.c"
+"lib/x86_vc/mmxidct.c"
+"lib/x86_vc/mmxstate.c"
+"lib/x86_vc/x86state.c"
+"win32/xmingw32/libtheoradec-all.def"
+)
+
+add_definitions(-D_USRDLL -DLIBTHEORA_EXPORTS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_BIND_TO_CURRENT_CRT_VERSION -DWIN32 -DNDEBUG -D_WINDOWS)
```
Comment (Orvid) [inline ports/libtheora/CMakeLists.txt]: You can drop the `_USRDLL`, `WIN32`, `NDEBUG`, and `_WINDOWS` defines, as CMake will add them when appropriate.

You'll also want to only define `LIBTHEORA_EXPORTS` if `BUILD_SHARED_LIBS` is true.

File: ports/libtheora/CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+
+add_library(theora ${SOURCES})
+target_link_libraries(theora ${LIBOGG_PATH})
```
Comment (Orvid) [inline ports/libtheora/CMakeLists.txt]: You need a call to `find_package` to find libogg before you can to this.

File: ports/libtheora/CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+"lib/x86_vc/mmxfrag.c"
+"lib/x86_vc/mmxidct.c"
+"lib/x86_vc/mmxstate.c"
+"lib/x86_vc/x86state.c"
+"win32/xmingw32/libtheoradec-all.def"
+)
+
+add_definitions(-D_USRDLL -DLIBTHEORA_EXPORTS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_NONSTDC_NO_DEPRECATE -D_BIND_TO_CURRENT_CRT_VERSION -DWIN32 -DNDEBUG -D_WINDOWS)
+
+add_library(theora ${SOURCES})
```
Comment (Orvid) [inline ports/libtheora/CMakeLists.txt]: It's not technically required, but it makes editing in Visual Studio much easier when you actually have the header files in the project, so add `${HEADERS}` to this list.

File: ports/libtheora/CONTROL [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Source: libtheora
+Version: 1.1.1
+Description: Theora is a free and open video compression format from the Xiph.org Foundation.
+#Build-Depends: libogg libvorbis
```
Comment (Orvid) [inline ports/libtheora/CONTROL]: You're linking against libogg in the CMake file, so you want to make sure this gets uncommented.

File: ports/libtheora/portfile.cmake [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+vcpkg_download_distfile(ARCHIVE
+    URLS "http://downloads.xiph.org/releases/theora/libtheora-1.1.1.tar.bz2"
+    FILENAME "libtheora-1.1.1.tar.bz2"
+    SHA512 9ab9b3af1c35d16a7d6d84f61f59ef3180132e30c27bdd7c0fa2683e0d00e2c791accbc7fd2c90718cc947d8bd10ee4a5940fb55f90f1fd7b0ed30583a47dbbd
+)
+vcpkg_extract_source_archive(${ARCHIVE})
+
+file(COPY ${CMAKE_CURRENT_LIST_DIR}/CMakeLists.txt DESTINATION ${SOURCE_PATH})
+
+file(TO_NATIVE_PATH "${VCPKG_ROOT_DIR}/installed/${TARGET_TRIPLET}/lib/ogg.lib" OGG_LIB_REL)
```
Comment (Orvid) [inline ports/libtheora/portfile.cmake]: The libogg port is built via CMake, so it would be better to link against its exported targets rather than doing this.

File: ports/libtheora/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+target_link_libraries(theora ${LIBOGG_PATH})
+
+install(FILES ${HEADERS} DESTINATION include/theora)
+
+install(TARGETS theora
+  RUNTIME DESTINATION bin
+  LIBRARY DESTINATION lib
+  ARCHIVE DESTINATION lib
+)
```
Comment (Orvid) [inline ports/libtheora/CMakeLists.txt]: Export the `theora` target.

Comment (vlj) [inline ports/libtheora/CMakeLists.txt]: what do you mean ?

**Comment by ras0219-msft** [comment]
Thanks @Orvid for reviewing this, and very thanks to @vlj for the PR and fixing the issues 👍.

It would be great in the future if we can get `ogg` to export cmake targets and use those directly. @Orvid, could you make a PR with changes to export the `theora` target so we can see a solid example?

**Comment by Orvid** [comment]
Sorry it's taken me a bit to respond, I've not been feeling great.

I would open a PR to make the changes to export the `theora` target if I could get my employer to sign the CLA, but apparently our legal team wants some changes that your legal team is hesitant to make. I don't know the details of it, and shall claim that both legal teams are just doing their jobs, so there's not much I can do about it. To the best of my knowledge, discussions are still ongoing. If I don't get confirmation back that they've reached an agreement in the next week, I'll try to kick off the process of getting a public fork of vcpkg setup instead, which should take far less work from legal, so that I can at least make the relatively well behaved portfiles I currently have at least available as reference for those working on new portfiles or trying to fix existing portfiles. I've had working portfiles for LibEvent and pthread, as well as a fixed version of double-conversion using CMake for the last couple of weeks, but haven't been able to submit them yet due to the CLA :(

TLDR; Until I can get my employer to sign the CLA, I'm limited to contributing feedback and (very) short snippets of code as feedback to PR authors :(

**Comment by ras0219-msft** [comment]
That's really unfortunate, but thank you for those contributions you've made so far nonetheless!
